common/src/main/java/org/keycloak/common/crypto/PemUtils.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2016 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.common.crypto;
+
+import java.security.Key;
+import java.security.NoSuchAlgorithmException;
+import java.security.PrivateKey;
+import java.security.PublicKey;
+import java.security.cert.Certificate;
+import java.security.cert.X509Certificate;
+
+/**
+ * Utility classes to extract PublicKey, PrivateKey, and X509Certificate from openssl generated PEM files
+ *
+ * @author <a href="mailto:bill@burkecentral.com">Bill Burke</a>
+ * @version $Revision: 1 $
+ */
+public interface PemUtils {
```

Comment:
2 comments to this:

- Is it possible to rename this to `PemUtilsProvider` ? See the similar comment like for `CertificateUtils` 

- I see that current class `org.keycloak.common.util.PemUtils` has only minimal usage of BouncyCastle. It is used only in the private method `private static String encode(Object obj)` - https://github.com/keycloak/keycloak/blob/main/common/src/main/java/org/keycloak/common/util/PemUtils.java#L139 . So I suggest to extract to cryptoProvider just the parts dependent on BouncyCastle, but keep the other parts in the `org.keycloak.common.util.PemUtils` to avoid some unecessary code duplicities.
 
So for example interface `PemUtilsProvider` can have only single method `String encode(Object obj)` . And most of the content of the class `org.keycloak.common.util.PemUtils` will be kept as is in the current Keycloak main. With only exception that private method `String encode(Object obj)` will be removed from this class and instead all the places of `org.keycloak.common.util.PemUtils`, which calls this private method, would instead call `cryptoProvider.getPemUtils().encode()` . Will this approach work for you?

Reply:
moved code to abstract class